### PR TITLE
Fix Callback exception when a removed task is the last one in the task instance list

### PR DESCRIPTION
**Pull request by joaopamaral**
When a task is removed and is returned in the last position of `get_task_instances`, the callback request will fail because it will try to access this task, but it doesn't exist anymore.

- Airflow 2.7.1:
```
ERROR - Error executing DagCallbackRequest callback for file: /dags/daily.py
Traceback (most recent call last):
  File "airflow/dag_processing/processor.py", line 711, in execute_callbacks
    self._execute_dag_callbacks(dagbag, request, session)
  File "airflow/utils/session.py", line 74, in wrapper
    return func(*args, **kwargs)
  File "airflow/dag_processing/processor.py", line 745, in _execute_dag_callbacks
    dag.handle_callback(
  File "airflow/utils/session.py", line 74, in wrapper
    return func(*args, **kwargs)
  File "airflow/models/dag.py", line 1422, in handle_callback
    ti.task = self.get_task(ti.task_id)
  File "airflow/models/dag.py", line 2496, in get_task
    raise TaskNotFound(f"Task {task_id} not found")
airflow.exceptions.TaskNotFound: Task removed_task not found
```
> This exception was noticed in airflow 2.7.1 but the issue is still in the latest version

To fix it we are filtering out the removed tasks to avoid having a removed task in the last position again.

- After
```python
>>> tis = dailydag.get_task_instances()
>>> dailydag.get_task(tis[-1].task_id)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "airflow/models/dag.py", line 2496, in get_task
    raise TaskNotFound(f"Task {task_id} not found")
airflow.exceptions.TaskNotFound: Task removed_task not found
```

- Before
```python
>>> tis = [ti for ti in dailydag.get_task_instances() if ti.state != TaskInstanceState.REMOVED]
>>> dailydag.get_task(tis[-1].task_id)
<Task(EmptyOperator): existing_task>
```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -1466,6 +1466,9 @@ def fetch_callback(
             # context for the callback.
             if dag.partial:
                 tis = [ti for ti in tis if not ti.state == State.NONE]
+            # filter out removed tasks
+            tis = [ti for ti in tis if ti.state != TaskInstanceState.REMOVED]
+            print(f"tis: {tis}")
```
Comment (hussein-awala) [inline airflow/models/dag.py]: We don't use print in the code; if there is a need to log something, we use `logging` package, but I don't think there is a need here. If you added it for debugging purposes, this comment will be a reminder to remove it before merging the PR.
```suggestion
```

**Review by hussein-awala — COMMENTED** [review]
Could you add a test for your change?

File: airflow/models/dag.py (hunk 1 above)
Comment (joaopamaral) [reply airflow/models/dag.py]: It was only for debbug! I've removed it and added the test. Thanks

**Review by vincbeck — APPROVED** [review]

File: airflow/models/dag.py [hunk 2]
```diff
@@ -1466,6 +1466,8 @@ def fetch_callback(
             # context for the callback.
             if dag.partial:
                 tis = [ti for ti in tis if not ti.state == State.NONE]
+            # filter out removed tasks
+            tis = [ti for ti in tis if ti.state != TaskInstanceState.REMOVED]
```
Comment (ephraimbuddy) [inline airflow/models/dag.py:1470]: We should merge this with line 1468 above

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]

File: airflow/models/dag.py (hunk 2 above)
Comment (joaopamaral) [reply airflow/models/dag.py:1470]: Done! 👍

**Comment by eladkal** [comment]
Tests are failing
Can you look into it @joaopamaral ?

File: airflow/models/dag.py [hunk 3]
```diff
@@ -1460,12 +1460,16 @@ def fetch_callback(
...
             # since dag.partial_subset is a subset of the dag.
             # This ensures that we will only use the accessible TI
             # context for the callback.
-            if dag.partial:
-                tis = [ti for ti in tis if not ti.state == State.NONE]
+            tis = [
+                ti
+                for ti in tis
+                if ti.state != TaskInstanceState.REMOVED or (dag.partial and not ti.state == State.NONE)
+            ]
```
Comment (ephraimbuddy) [inline airflow/models/dag.py]: What you had earlier seems better, can we go back to it? Sorry for pushing you to this direction

**Review by ephraimbuddy — APPROVED** [review]

**Pull request merged**